### PR TITLE
Support MySQL and postgresql test (In progress)

### PR DESCRIPTION
**Pull request by ttt-77**
- [x] Add tests for MySQL and PostgreSQL databases
- [x]  Add engine.dispose() to prevent potential warnings and runtime errors.
- [x] Resolve the "Data too long for column 'entity'" issue in MySQL.
- [x]  Resolve the IntegrityError as outlined in issue #28

### CODE REVIEW

File: aidb/engine/approx_aggregate_engine.py [hunk 1]
```diff
@@ -161,12 +165,12 @@ async def execute_inference_services(self, query: Query, sample_df: pd.DataFrame
...
       )
 
-      input_df = await conn.run_sync(lambda conn: pd.read_sql(text(query_add_filter_key.sql_str), conn))
+      input_df = await conn.run_sync(lambda conn: pd.read_sql_query(text(query_add_filter_key.sql_str), conn))
       await bound_service.infer(input_df)
       inference_services_executed.add(bound_service.service.name)
 
 
-  async def get_results_for_each_blob(self, query: Query, sample_df: pd.DataFrame, conn) -> pd.DataFrame:
+  async def get_results_for_each_blob(self, query: Query, sample_df: pd.DataFrame) -> pd.DataFrame:
```
Comment (ttt-77) [inline aidb/engine/approx_aggregate_engine.py:173]: Restart the engine within the function; there's no need to pass 'conn'.

File: aidb/inference/bound_inference_service.py [hunk 2]
```diff
@@ -64,21 +64,21 @@ def get_table(conn: sqlalchemy.engine.base.Connection):
...
-        logger.debug('New table col name', new_table_col_name)
-        logger.debug('Ref column', str(column))
-        fk_ref_table_name = str(column).split('.')[0]
-        if fk_ref_table_name not in fk_constraints:
-          fk_constraints[fk_ref_table_name] = {'cols': [], 'cols_refs': []}
-        # both tables will have same column name
-        fk_constraints[fk_ref_table_name]['cols'].append(new_table_col_name)
-        fk_constraints[fk_ref_table_name]['cols_refs'].append(column)
-        columns.append(sqlalchemy.schema.Column(new_table_col_name, column.type))
+        if column.primary_key:
```
Comment (ttt-77) [inline aidb/inference/bound_inference_service.py:67]: For the cache table, store only the primary key to avoid foreign key constraint errors.

File: tests/db_utils/db_setup.py [hunk 3]
```diff
@@ -132,49 +144,47 @@ def get_insertion_order(conn, gt_csv_files):
...
-async def setup_config_tables(engine):
+async def setup_config_tables(conn):
   def create_blob_metadata_table(conn):
     Base = declarative_base()
 
     class BlobTables(Base):
       __tablename__ = BLOB_TABLE_NAMES_TABLE
-      table_name = sqlalchemy.Column(sqlalchemy.String, primary_key=True)
-      blob_key = sqlalchemy.Column(sqlalchemy.String, primary_key=True)
+      table_name = sqlalchemy.Column(sqlalchemy.String(30), primary_key=True)
```
Comment (ttt-77) [inline tests/db_utils/db_setup.py:186]: String length is required for MySQL database

File: aidb/engine/tasti_engine.py [hunk 4]
```diff
@@ -95,7 +95,8 @@ def get_score_query_str(self, query: Query, rep_table_name: str) -> (str, List[L
     score_list = []
     # FIXME: for different database, the IN grammar maybe different
     for fp in filering_predicate_score_map:
-      score_list.append(f'IIF({fp}, 1, 0) AS {filering_predicate_score_map[fp]}')
+      # score_list.append(f'IIF({fp}, 1, 0) AS {filering_predicate_score_map[fp]}')
+      score_list.append(f'(CASE WHEN {fp} THEN 1 ELSE 0 END) AS {filering_predicate_score_map[fp]}')
```
Comment (ttt-77) [inline aidb/engine/tasti_engine.py:99]: 'IIF' function doesn't exist in either MySQL or PostgreSQL databases.

File: tests/db_utils/db_setup.py [hunk 5]
```diff
@@ -132,49 +144,46 @@ def get_insertion_order(conn, gt_csv_files):
...
-        continue
-      df = pd.read_csv(csv_fname)
-      for column in df.columns:
-        column_info = extract_column_info(table_name, column)
-        df.rename(columns={column: column_info.name}, inplace=True)
 
-      if table_name.startswith('mapping'):
-        _, _, blob_mapping_table_name = table_name_for_rep_and_topk_and_blob_mapping(['blobs_00'])
-        table_name = blob_mapping_table_name
-      # FIXME: in case of mysql, this function doesn't wait, hence throwing integrity error
```
Comment (ttt-77) [inline tests/db_utils/db_setup.py:153]: "User
pymysql.err.IntegrityError: (1452, 'Cannot add or update a child row: a foreign key constraint fails (`aidb_gt_8000`.`counts03`, CONSTRAINT `counts03_ibfk_1` FOREIGN KEY (`frame`) REFERENCES `blobs_00` (`frame`))')"
I encountered an Integrity Error due to setting autoincrement=True for a primary column. In the MySQL database, this setting caused the value 0 in the 'frame' column to change to 1. After setting autoincrement=False for this column, the error no longer occurred.

**Comment by ttt-77** [comment]
Could you please review this PR? @ddkang

File: aidb/engine/approx_aggregate_engine.py [hunk 6]
```diff
@@ -161,12 +165,12 @@ async def execute_inference_services(self, query: Query, sample_df: pd.DataFrame
           new_query
       )
 
-      input_df = await conn.run_sync(lambda conn: pd.read_sql(text(query_add_filter_key.sql_str), conn))
+      input_df = await conn.run_sync(lambda conn: pd.read_sql_query(text(query_add_filter_key.sql_str), conn))
```
Comment (ddkang) [inline aidb/engine/approx_aggregate_engine.py:168]: What's the difference here?

File: aidb/engine/approx_aggregate_engine.py (hunk 1 above)
Comment (ddkang) [reply aidb/engine/approx_aggregate_engine.py:173]: Why do we need this change? Is it possible to separate this out into a separate PR?

File: aidb/engine/approx_select_engine.py [hunk 7]
```diff
@@ -76,22 +76,25 @@ async def get_inference_results(self, query:Query, sampled_index: List[int], con
         query_no_where
     )
 
-    all_df = await conn.run_sync(lambda conn: pd.read_sql(text(all_query_add_filter_key.sql_str), conn))
-    # drop duplicated columns, this will happen when 'select *'
-    all_df = all_df.loc[:, ~all_df.columns.duplicated()]
-    all_df.set_index(VECTOR_ID_COLUMN, inplace=True, drop=True)
+    async with self._sql_engine.begin() as conn:
```
Comment (ddkang) [inline aidb/engine/approx_select_engine.py:79]: Why is this changed?

File: aidb/engine/tasti_engine.py (hunk 4 above)
Comment (ddkang) [reply aidb/engine/tasti_engine.py:97]: Does this also work on sqlite?

File: tests/db_utils/db_setup.py [hunk 8]
```diff
@@ -69,49 +71,59 @@ async def setup_db(db_url: str, db_name: str, data_dir: str):
...
-      multi_table_fk_constraints = []
-      for tbl, fk_cons in fk_constraints.items():
-        multi_table_fk_constraints.append(ForeignKeyConstraint(fk_cons['cols'], fk_cons['cols_refs']))
-
-      _ = sqlalchemy.Table(table_name, metadata, *[
-        sqlalchemy.Column(c_info.name, c_info.dtype, primary_key=c_info.is_primary_key) for c_info in columns_info
-      ], *multi_table_fk_constraints)
-
-    await conn.run_sync(lambda conn: metadata.create_all(conn))
+  try:
```
Comment (ddkang) [inline tests/db_utils/db_setup.py:74]: What are the major differences here?

File: tests/db_utils/db_setup.py (hunk 3 above)
Comment (ddkang) [reply tests/db_utils/db_setup.py:185]: Make this longer - lots of automatically generated schemas have long names, like 1024

File: aidb/engine/tasti_engine.py (hunk 4 above)
Comment (ttt-77) [reply aidb/engine/tasti_engine.py:97]: Yes

File: aidb/engine/approx_aggregate_engine.py (hunk 6 above)
Comment (ttt-77) [reply aidb/engine/approx_aggregate_engine.py:168]: 'read_sql' is a convenience wrapper around read_sql_table and read_sql_query.
In PostgreSQL, using read_sql can lead to an sqlalchemy.exc.DBAPIError. I think the driver of PostgreSQL might have compatibility issues with 'read_sql'.

File: aidb/engine/approx_aggregate_engine.py (hunk 1 above)
Comment (ttt-77) [reply aidb/engine/approx_aggregate_engine.py:173]: In MySQL database, after running inference (there is another engine.begin() in 'infer()' function), the database will lose connection. We need to restart the engine, which is implemented inside 'get_results_for_each_blob'. So there's no need to pass 'conn'.

File: aidb/engine/approx_select_engine.py (hunk 7 above)
Comment (ttt-77) [reply aidb/engine/approx_select_engine.py:79]: Similar to the issue encountered with approx_aggregation in MySQL database. I will separate them out.

File: tests/db_utils/db_setup.py [hunk 9]
```diff
@@ -69,49 +71,59 @@ async def setup_db(db_url: str, db_name: str, data_dir: str):
...
+          column_info = extract_column_info(table_name, column)
+          dtype = python_type_to_sqlalchemy_type(df[column].dtype)
+          if dtype == sqlalchemy.String:
+            # TODO: VAR CHAR lenth should be based on the number of characters
+            if column_info.is_primary_key or column_info.refers_to:
+              column_info.dtype = sqlalchemy.String(20)
+            else:
+              column_info.dtype = sqlalchemy.TEXT
+          else:
+            column_info.dtype = dtype
```
Comment (ttt-77) [inline tests/db_utils/db_setup.py:95]: In a MySQL database, using sqlalchemy.String(20) can lead to a "Data too long for column 'entity'" error due to the character limit. In addition, using sqlalchemy.TEXT for a primary key column results in the error "BLOB/TEXT column 'tweet_id' used in key specification without a key length," indicating that the primary key column requires a column type with a fixed length.

File: tests/db_utils/db_setup.py (hunk 8 above)
Comment (ttt-77) [reply tests/db_utils/db_setup.py:74]: 1. Add 'engine.dispose()' in the 'finally' block to prevent a potential "ResourceWarning: Unclosed connection".  
2. Revise the definition of the column type in the MySQL database; relevant columns are detailed below.

File: aidb/engine/approx_aggregate_engine.py (hunk 6 above)
Comment (ddkang) [reply aidb/engine/approx_aggregate_engine.py:168]: Can we move this out to a separate PR? Would be much easier to review

File: tests/db_utils/db_setup.py (hunk 9 above)
Comment (ddkang) [reply tests/db_utils/db_setup.py:95]: Wait why can't we have strings of length 20? What about varchar?

File: tests/db_utils/db_setup.py [hunk 10]
```diff
@@ -69,49 +71,59 @@ async def setup_db(db_url: str, db_name: str, data_dir: str):
...
+              f"{column_info.refers_to[0]}.{column_info.refers_to[1]}")
+
+        multi_table_fk_constraints = []
+        for tbl, fk_cons in fk_constraints.items():
+          multi_table_fk_constraints.append(ForeignKeyConstraint(fk_cons['cols'], fk_cons['cols_refs']))
+
+        _ = sqlalchemy.Table(
+            table_name,
+            metadata,
+            *[sqlalchemy.Column(c_info.name, c_info.dtype, primary_key=c_info.is_primary_key, autoincrement=False)
```
Comment (ttt-77) [inline tests/db_utils/db_setup.py:115]: Configure the 'autoincrement' attribute to 'False' to prevent the unintended conversion of '0' to '1' in MySQL database, which can lead to an integrity error.

File: tests/db_utils/db_setup.py (hunk 3 above)
Comment (ttt-77) [reply tests/db_utils/db_setup.py]: For the primary key column, we can only set the maximum length up to 256

File: tests/db_utils/db_setup.py (hunk 9 above)
Comment (ttt-77) [reply tests/db_utils/db_setup.py]: For the primary key column, I have changed it to 256. For other columns, should we use data type VARCHAR with a limit length or TEXT?

Comment (ddkang) [reply tests/db_utils/db_setup.py]: Let's do 2048, I think that works for nearly all cases

**Review by ddkang — APPROVED** [review]
